modules/infra/control/port.c

```diff
@@ -457,6 +463,72 @@ static struct iface_type iface_type_port = {
 	.to_api = port_to_api,
 };
 
+static int mcast_count(struct iface_info_port *p) {
+	int count = 0;
+	for (int i = 0; i < GR_PORT_MCAST_ADDR_MAX; i++) {
+		if (p->mcast_refcnt[i])
+			count++;
+	}
+	return count;
+}
+
+static void mcast_sort(struct iface_info_port *p) {
+	for (int i = 0; i < GR_PORT_MCAST_ADDR_MAX - 1; i++) {
+		if (p->mcast_refcnt[i] == 0) {
+			for (int j = i + 1; i < GR_PORT_MCAST_ADDR_MAX; j++) {
+				if (p->mcast_refcnt[j]) {
+					p->mcast_refcnt[i] = p->mcast_refcnt[j];
+					memcpy(&p->mcast[i],
+					       &p->mcast[j],
+					       sizeof(struct rte_ether_addr));
+					memset(&p->mcast[j], 0, sizeof(struct rte_ether_addr));
+					break;
+				}
+			}
+		}
+	}
+}
+
+int gr_port_mcast_addr_add(struct iface_info_port *p, struct rte_ether_addr *mcast_addr) {
+	int i;
+	for (i = 0; i < GR_PORT_MCAST_ADDR_MAX; i++) {
+		if (rte_is_same_ether_addr(mcast_addr, &p->mcast[i])) {
+			LOG(DEBUG, "multicast address already filtered by port %d", p->port_id);
+			p->mcast_refcnt[i]++;
```

Comment:
In which case do we have multiple modules requiring the same mcast address?

Reply:
Multiple protocols may use the same multicast mac address, or the same multicast IPs from 2 VRFs.